filebeat/input/v2/input-cursor/input.go

```diff
@@ -73,7 +103,73 @@ func (inp *managedInput) Run(
 	ctx input.Context,
 	pipeline beat.PipelineConnector,
 ) (err error) {
-	panic("TODO: implement me")
+	// Setup cancellation using a custom cancel context. All workers will be
+	// stopped if one failed badly by returning an error.
+	cancelCtx, cancel := context.WithCancel(ctxtool.FromCanceller(ctx.Cancelation))
+	defer cancel()
+	ctx.Cancelation = cancelCtx
+
+	var grp unison.MultiErrGroup
+	for _, source := range inp.sources {
+		source := source
+		grp.Go(func() (err error) {
+			// refine per worker context
+			inpCtx := ctx
+			inpCtx.ID = ctx.ID + "::" + source.Name()
+			inpCtx.Logger = ctx.Logger.With("source", source.Name())
+
+			if err = inp.runSource(inpCtx, inp.manager.store, source, pipeline); err != nil {
+				cancel()
+			}
+			return err
+		})
+	}
+
+	if errs := grp.Wait(); len(errs) > 0 {
+		return sderr.WrapAll(errs, "input %{id} failed", ctx.ID)
+	}
+	return nil
+}
+
+func (inp *managedInput) runSource(
+	ctx input.Context,
+	store *store,
+	source Source,
+	pipeline beat.PipelineConnector,
+) (err error) {
+	// Setup error recovery/reporting
```

Comment:
Nit: this comment is unnecessary.

Reply:
Will remove. They are leftovers from prototyping and having a `panic` implementation